# Исследование базы данных крупного сервиса для чтения книг по подписке

В этом исследовании проанализируем базу данных крупного сервиса для чтения книг по подписке.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Общая-информация-о-файлах" data-toc-modified-id="Общая-информация-о-файлах-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Общая информация о файлах</a></span></li><li><span><a href="#Исследовательский-анализ-данных" data-toc-modified-id="Исследовательский-анализ-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследовательский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Посчитаем,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитаем,-сколько-книг-вышло-после-1-января-2000-года-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Посчитаем, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитаем-количество-обзоров-и-среднюю-оценку-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Для каждой книги посчитаем количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определим-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Определим издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определим-автора-с-самой-высокой-средней-оценкой-книг-(учитываем-только-книги-с-50-и-более-оценками)" data-toc-modified-id="Определим-автора-с-самой-высокой-средней-оценкой-книг-(учитываем-только-книги-с-50-и-более-оценками)-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Определим автора с самой высокой средней оценкой книг (учитываем только книги с 50 и более оценками)</a></span></li><li><span><a href="#Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитаем-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

## Общая информация о файлах

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

#добавим функцию для вывода запросов SQL
def read (query):
    display(pd.io.sql.read_sql(sql=text(query), con = engine))

In [2]:
# выведем первые 5 строк таблицы books
query = '''SELECT * FROM books LIMIT 5'''

read (query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
# выведем первые 5 строк таблицы authors
query = '''SELECT * FROM authors LIMIT 5'''

read (query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
# выведем первые 5 строк таблицы publishers
query = '''SELECT * FROM publishers LIMIT 5'''

read (query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
# выведем первые 5 строк таблицы ratings
query = '''SELECT * FROM ratings LIMIT 5'''

read (query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
# выведем первые 5 строк таблицы reviews
query = '''SELECT * FROM reviews LIMIT 5'''

read (query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Исследовательский анализ данных

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [7]:
query = '''
SELECT COUNT(title) 
FROM books
WHERE publication_date > '2000-01-01'
'''
read (query)

count
0    819

После 1 января 2000 года вышло 819 книг и брошюр.

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [8]:
query = '''
SELECT DISTINCT b.title,
       SUM(r.review_id),
       ROUND(AVG(rat.rating),2)
FROM books AS b
JOIN reviews AS r ON r.book_id = b.book_id
JOIN ratings AS rat ON rat.book_id = b.book_id
GROUP BY 1
ORDER BY 2 DESC
'''

read (query)

title      sum  round
0                              Twilight (Twilight  #1)  2963520   3.66
1                  The Hobbit  or There and Back Again  1101672   4.13
2                               The Catcher in the Rye   956406   3.83
3    The Lightning Thief (Percy Jackson and the Oly...   805938   4.08
4    The Fellowship of the Ring (The Lord of the Ri...   738520   4.39
..                                                 ...      ...    ...
988     A Christmas Carol and Other Christmas Writings       74   4.50
989  13 Little Blue Envelopes (Little Blue Envelope...       45   4.67
990  1491: New Revelations of the Americas Before C...       30   4.50
991                                       'Salem's Lot        9   3.67
992                 1 000 Places to See Before You Die        6   2.50

[993 rows x 3 columns]

Всего 993 книги с уникальным названием. 

Больше всего обзоров выпущено на книгу Twilight (Сумерки) 1 часть, однако средняя оценка  критиков довольно низкая: всего 3.66.

А меньше всего обзоров выпущено на книгу 1 000 Places to See Before You Die (1000 мест, которые стоит увидеть перед смертью), и средняя оценка еще ниже: 2.5.

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query = '''
SELECT us.publisher
FROM (
      SELECT DISTINCT p.publisher,
             COUNT(b.book_id)
      FROM books AS b
      JOIN publishers AS p ON p.publisher_id = b.publisher_id
      WHERE num_pages > '50'
      GROUP BY 1
      ORDER BY 2 DESC
      LIMIT 1
      ) us
'''

read (query)

publisher
0  Penguin Books

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books.

Penguin Books — британское издательство, основанное в 1935 году в Лондоне сэром Алленом Лейном и его братьями Ричардом и Джоном. Главной заслугой издательства считают демократизацию книжного рынка, превратившую книгу из предмета роскоши в удовольствие, доступное массам.

### Определим автора с самой высокой средней оценкой книг (учитываем только книги с 50 и более оценками)

In [10]:
query = '''

WITH new AS (--отберем книги с 50 и более оценками
             SELECT DISTINCT book_id,
                    COUNT(rating_id) AS count,
                    AVG(rating) AS avg
             FROM ratings AS rat
             GROUP BY 1             
             HAVING COUNT(rating_id) >= '50'
             ORDER BY 3 DESC
            ) 
SELECT DISTINCT author,
       title,
       avg
FROM new AS new
INNER JOIN books AS b ON new.book_id = b.book_id
INNER JOIN authors AS a ON b.author_id = a.author_id
LIMIT 1

'''

read (query)

author  \
0  J.K. Rowling/Mary GrandPré   

                                               title     avg  
0  Harry Potter and the Chamber of Secrets (Harry...  4.2875

Книга с 50 и более оценками, получившая самый высокий средний рейтинг: Гарри Поттер и Тайная комната (Harry Potter and the Chamber of Secrets).

Ее авторами являются:
- Джоа́н Ро́улинг, известная под псевдонимами Дж. К. Роулинг, Джоан Кэ́тлин Роулинг и Ро́берт Гелбре́йт (британская писательница, сценаристка и кинопродюсер, наиболее известная как автор серии романов о Гарри Поттере)
- Мэри ГрандПре (американский иллюстратор, наиболее известная своими иллюстрациями обложек и глав книг о Гарри Поттере в их американских изданиях, опубликованных Scholastic).

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [11]:
query = '''
WITH table_1 AS (SELECT username,
                        COUNT(rating_id) AS count_rating
                 FROM ratings
                 GROUP BY 1
                ),
table_2 AS (SELECT username,
                        COUNT(review_id) AS count_review
                 FROM reviews 
                 GROUP BY 1
                 ORDER BY 2 DESC
                )
SELECT ROUND(AVG(t_2.count_review))
FROM table_1 AS t_1
INNER JOIN table_2 AS t_2 ON t_1.username = t_2.username
WHERE count_rating > 50
'''

read (query)

round
0   24.0

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок, составило 24.

## Вывод

Итак, проанализировав базу данных крупного сервиса для чтения книг по подписке, можно выделить следующее:
1. После 1 января 2000 года вышло 819 книг и брошюр.
2. Всего 993 книги с уникальным названием.
3. Больше всего обзоров выпущено на книгу Twilight (Сумерки) 1 часть, однако средняя оценка критиков довольно низкая: всего 3.66.
4. Меньше всего обзоров выпущено на книгу 1 000 Places to See Before You Die (1000 мест, которые стоит увидеть перед смертью), и средняя оценка еще ниже: 2.5.
5. Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books (британское издательство, основанное в 1935 году в Лондоне сэром Алленом Лейном и его братьями Ричардом и Джоном, главной заслугой которого считают демократизацию книжного рынка, превратившую книгу из предмета роскоши в удовольствие, доступное массам).
6. Книга с 50 и более оценками, получившая самый высокий средний рейтинг: Гарри Поттер и Тайная комната (Harry Potter and the Chamber of Secrets).
7. Авторами книги с 50 и более оценками, получившей самый высокий средний рейтинг, являются:
- Джоа́н Ро́улинг, известная под псевдонимами Дж. К. Роулинг, Джоан Кэ́тлин Роулинг и Ро́берт Гелбре́йт (британская писательница, сценаристка и кинопродюсер, наиболее известная как автор серии романов о Гарри Поттере)
- Мэри ГрандПре (американский иллюстратор, наиболее известная своими иллюстрациями обложек и глав книг о Гарри Поттере в их американских изданиях, опубликованных Scholastic).
7. Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок, составило 24.